modyn/supervisor/internal/pipeline_executor/pipeline_executor.py

```diff
@@ -127,6 +133,64 @@ def _persist_pipeline_log(self) -> None:
             json.dump(self.pipeline_log, logfile, indent=4)
 
+    # pylint: disable=too-many-locals
+    def _start_evaluations(
+        self, trigger_id: int, model_id: int, trigger_set_first_timestamp: int, trigger_set_last_timestamp: int
+    ) -> None:
+        assert self.grpc.evaluator is not None, "Evaluator not initialized."
+        eval_dataset_config = self.pipeline_config["evaluation"]["dataset"]
+        eval_dataset_id = eval_dataset_config["dataset_id"]
+        eval_strategy_config = self.pipeline_config["evaluation"]["eval_strategy"]
+        eval_strategy_module = dynamic_module_import("modyn.supervisor.internal.eval_strategies")
+        eval_strategy: AbstractEvalStrategy = getattr(eval_strategy_module, eval_strategy_config["name"])(
+            eval_strategy_config["config"]
+        )
+        if EVALUATION_RESULTS not in self.pipeline_log:
+            self.pipeline_log[EVALUATION_RESULTS] = {}
+
+        self.pipeline_log[EVALUATION_RESULTS][trigger_id] = {}
+
+        for interval_start, interval_end in eval_strategy.get_eval_intervals(
+            first_timestamp=trigger_set_first_timestamp,
+            last_timestamp=trigger_set_last_timestamp,
+        ):
+            logger.info(f"Evaluation Starts for model {model_id} on split {interval_start} to {interval_end}.")
+            request = GRPCHandler.prepare_evaluation_request(
+                eval_dataset_config,
+                model_id,
+                self.pipeline_config["evaluation"]["device"],
+                interval_start,
+                interval_end,
+            )
+            self._sw.start("evaluate_model", overwrite=True)
+            response: EvaluateModelResponse = self.grpc.evaluator.evaluate_model(request)
+            interval_name = f"{interval_start}-{interval_end}"
+            if not response.evaluation_started:
+                reason = EvaluationAbortedReason.DESCRIPTOR.values_by_number[response.eval_aborted_reason].name
+                logger.error(
+                    f"Evaluation for model {model_id} on split {interval_start} to {interval_end} not started with "
+                    f"reason: {reason}."
+                )
+                self.pipeline_log[EVALUATION_RESULTS][trigger_id][interval_name] = {
+                    "failure_reason": reason,
+                }
+                self._sw.stop("evaluate_model")
+            else:
+                logger.info(f"Evaluation started for model {model_id} on split {interval_start} to {interval_end}.")
+                reporter = EvaluationStatusReporter(
+                    self.eval_status_queue, response.evaluation_id, eval_dataset_id, response.dataset_size
+                )
+                reporter.create_tracker()
+                evaluation = {response.evaluation_id: reporter}
+                assert self.current_training_id is not None, "Training ID not set."
+                self.grpc.wait_for_evaluation_completion(self.current_training_id, evaluation)
```

Comment:
So right now we cannot do parallel evaluation, correct? We always wait for one evaluation to start before starting the next one?

One way to do it would be to implement parallel evaluation at the evaluator itself, where it has a queue of incoming requests, executing as much in parallel as possible, and the pipeline executor will just submit a bunch of requests and wait for all of them to finish.

Reply:
anyways, not relevant for this pr but let's keep this in mind for the future

Reply:
Yes now it's one after another. And we should measure if we can get speed up by executing multiple eval jobs in parallel (and I doubt it), because as far as I know, if we don't enable [Nvidia Multi-Process Service](https://docs.nvidia.com/deploy/mps/index.html), it's still executed one after another on GPU, or in best case time-sliced ([source](https://stackoverflow.com/questions/31643570/running-more-than-one-cuda-applications-on-one-gpu)).

For this PR let's do it one-after-another.

Reply:
hmm, okay, I will also chat with Foteini about this. Thanks for the pointers!